src/paths.md

```diff
@@ -40,42 +40,34 @@ mod m {
 
 > **<sup>Syntax</sup>**\
 > _PathInExpression_ :\
-> &nbsp;&nbsp; `::`<sup>?</sup> _PathExprSegment_ (`::` _PathExprSegment_)<sup>\*</sup>
+> &nbsp;&nbsp; `::`<sup>?</sup> _PathExprSegment_ ( `::` _PathExprSegment_ )<sup>\*</sup>
 >
 > _PathExprSegment_ :\
-> &nbsp;&nbsp; _PathIdentSegment_ (`::` _GenericArgs_)<sup>?</sup>
+> &nbsp;&nbsp; _PathIdentSegment_ ( `::` `<` _GenericArguments_<sup>?</sup> `>` )<sup>?</sup>
 >
 > _PathIdentSegment_ :\
 > &nbsp;&nbsp; [IDENTIFIER] | `super` | `self` | `Self` | `crate` | `$crate`
 >
-> _GenericArgs_ :\
-> &nbsp;&nbsp; &nbsp;&nbsp; `<` `>`\
-> &nbsp;&nbsp; | `<` _GenericArgsLifetimes_ `,`<sup>?</sup> `>`\
-> &nbsp;&nbsp; | `<` _GenericArgsTypes_ `,`<sup>?</sup> `>`\
-> &nbsp;&nbsp; | `<` _GenericArgsBindings_ `,`<sup>?</sup> `>`\
-> &nbsp;&nbsp; | `<` _GenericArgsTypes_ `,` _GenericArgsBindings_ `,`<sup>?</sup> `>`\
-> &nbsp;&nbsp; | `<` _GenericArgsLifetimes_ `,` _GenericArgsTypes_ `,`<sup>?</sup> `>`\
-> &nbsp;&nbsp; | `<` _GenericArgsLifetimes_ `,` _GenericArgsBindings_ `,`<sup>?</sup> `>`\
-> &nbsp;&nbsp; | `<` _GenericArgsLifetimes_ `,` _GenericArgsTypes_ `,` _GenericArgsBindings_ `,`<sup>?</sup> `>`
+> _GenericArguments_ :\
+> &nbsp;&nbsp; _GenericArgument_ ( `,` _GenericArgument_ )<sup>\*</sup> `,`<sup>?</sup>
 >
-> _GenericArgsLifetimes_ :\
-> &nbsp;&nbsp; [_Lifetime_] (`,` [_Lifetime_])<sup>\*</sup>
->
-> _GenericArgsTypes_ :\
-> &nbsp;&nbsp; [_Type_] (`,` [_Type_])<sup>\*</sup>
->
-> _GenericArgsBindings_ :\
-> &nbsp;&nbsp; _GenericArgsBinding_ (`,` _GenericArgsBinding_)<sup>\*</sup>
->
-> _GenericArgsBinding_ :\
-> &nbsp;&nbsp; [IDENTIFIER] `=` [_Type_]
+> _GenericArgument_ :\
+> &nbsp;&nbsp; &nbsp;&nbsp; [_Lifetime_]\
+> &nbsp;&nbsp; | [_Type_]\
+> &nbsp;&nbsp; | [IDENTIFIER] `=` [_Type_]
 
 Paths in expressions allow for paths with generic arguments to be specified. They are
 used in various places in [expressions] and [patterns].
 
 The `::` token is required before the opening `<` for generic arguments to avoid
 ambiguity with the less-than operator. This is colloquially known as "turbofish" syntax.
 
+Generic arguments are passed in the order that they're declared in. In particular,
+lifetime arguments must come before type arguments. The last kind of argument,
+`Ident = Type`, is for specifying [associated types] in [trait bounds].
+These arguments are not positional (i.e. their order does not matter), and they
+must come after all the other (positional) arguments.
```

Comment:
The main point was that if you already know that the argument order matches the declaration, then the fact that lifetime argument come before type arguments directly follows from the fact that function, trait and method _declarations_ must put lifetime arguments first.

I’ll try to get rid of the contradictory statements, and reintroducing (at least some of) the names seems like a very good idea to make things clearer.

Reply:
> then the fact that lifetime argument come before type arguments directly follows from […]

The compiler [seems to (maybe) apply the same logic](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=71b461a4609bba73e4fb5ce3ebb9ac68) (note how it only complains about “associated type bindings” coming before “generic parameters” without knowing the details `Foo`).

Well, it can also generate [weird messages](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=87df0b30db94a750f4b70b8704b2cd4b) if you _do_ let it know of a `Foo`.